```

======================================================================
FAIL: custom.aaa.tests.test_eligible_couple_query_helper:TestEligibleCoupleBeneficiarySections.test_eligible_couple_details_with_history
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/custom/aaa/tests/test_eligible_couple_query_helper.py", line 71, in test_eligible_couple_details_with_history
    'preferredFamilyPlaningMethod': 'N/A',
AssertionError: {u'maleChildrenAlive': 0, u'preferredFamilyPlaningMethod': u'N/A', u'femaleChild [truncated]... != {u'maleChildrenAlive': 0, u'preferredFamilyPlaningMethod': u'N/A', u'femaleChild [truncated]...
  {u'ashaVisit': u'N/A',
   u'familyPlaningMethod': u'condom',
-  u'familyPlanningMethodDate': datetime.date(2019, 3, 1),
+  u'familyPlanningMethodDate': u'2019-03-01',
   u'femaleChildrenAlive': 0,
   u'femaleChildrenBorn': 0,
   u'maleChildrenAlive': 0,
   u'maleChildrenBorn': 0,
   u'preferredFamilyPlaningMethod': u'N/A',
   u'previousFamilyPlanningMethod': u'iud'}
```

--- a/custom/aaa/dbaccessors.py
+++ b/custom/aaa/dbaccessors.py
@@ -4,9 +4,11 @@ from collections import OrderedDict
 from datetime import date
 
 from dateutil.relativedelta import relativedelta
+from django.db import connections
 from django.db.models import Case, F, Func, IntegerField, Sum, When
 from django.db.models.functions import ExtractYear, ExtractMonth
 
+from corehq.sql_db.connections import get_aaa_db_alias
 from custom.aaa.models import (
     CcsRecord,
     Child,
@@ -224,44 +226,48 @@ class PregnantWomanQueryHelper(object):
 
     @classmethod
     def list(cls, domain, date_, location_filters, sort_column):
-        return (
-            Woman.objects.annotate(
-                age=ExtractYear(Func(F('dob'), function='age')),
-            ).filter(
-                domain=domain,
-                **location_filters
-            ).extra(
-                select={
-                    'highRiskPregnancy': '\'N/A\'',
-                    'noOfAncCheckUps': '\'N/A\'',
-                    'pregMonth': '\'N/A\'',
-                    'id': 'person_case_id',
-                },
-                where=["daterange(%s, %s) && any(pregnant_ranges)"],
-                params=[date_, date_ + relativedelta(months=1)]
-            ).values(
-                'id', 'name', 'age', 'pregMonth',
-                'highRiskPregnancy', 'noOfAncCheckUps'
-            ).order_by(sort_column)
-        )
+        location_query = ''
+        if location_filters:
+            location_query = [
+                "{loc} = %({loc})s".format(loc=loc) for loc in location_filters.keys()
+            ]
+            location_query = " AND ".join(location_query)
+            location_query = location_query + " AND"
 
-    @classmethod
-    def update_list(cls, data):
-        for beneficiary in data:
-            ccs_record = CcsRecord.objects.annotate(
-                pregMonth=ExtractMonth(Func(F('preg_reg_date'), function='age')),
-            ).filter(
-                person_case_id=beneficiary['id']
-            ).extra(
-                select={
-                    'highRiskPregnancy': 'hrp',
-                    'noOfAncCheckUps': 'num_anc_checkups',
-                }
-            ).values(
-                'highRiskPregnancy', 'noOfAncCheckUps', 'pregMonth'
-            ).first()
-            beneficiary.update(ccs_record)
-        return data
+        query, params = """
+            SELECT
+                (woman.person_case_id) AS "id",
+                woman.name AS "name",
+                EXTRACT('year' FROM age("woman"."dob")) AS "age",
+                EXTRACT('month' FROM age(ccs_record.preg_reg_date)) AS "pregMonth",
+                ccs_record.hrp AS "highRiskPregnancy",
+                ccs_record.num_anc_checkups AS "noOfAncCheckUps"
+            FROM "{woman_table}" woman
+            JOIN "{ccs_record_table}" ccs_record ON ccs_record.person_case_id=woman.person_case_id
+            WHERE (
+                woman.domain = %(domain)s AND
+                {location_where}
+                (daterange(%(start_date)s, %(end_date)s) && ANY(pregnant_ranges))
+            ) ORDER BY {sort_col}
+        """.format(
+            location_where=location_query,
+            woman_table=Woman._meta.db_table,
+            ccs_record_table=CcsRecord._meta.db_table,
+            sort_col=sort_column
+        ), {
+            'domain': domain,
+            'start_date': date_,
+            'end_date': date_ + relativedelta(months=1),
+        }
+        params.update(location_filters)
+        db_alias = get_aaa_db_alias()
+        with connections[db_alias].cursor() as cursor:
+            cursor.execute(query, params)
+            desc = cursor.description
+            return [
+                dict(zip([col[0] for col in desc], row))
+                for row in cursor.fetchall()
+            ]
 
     def pregnancy_details(self):
         data = CcsRecord.objects.extra(

--- a/custom/aaa/static/aaa/js/models/pregnant_women.js
+++ b/custom/aaa/static/aaa/js/models/pregnant_women.js
@@ -37,6 +37,10 @@ hqDefine("aaa/js/models/pregnant_women", [
         self.highRiskPregnancy = ko.computed(function () {
             return self.highRiskPregnancy() === 'yes' ? 'Yes' : 'No';
         });
+
+        self.noOfAncCheckUps = ko.computed(function () {
+            return self.noOfAncCheckUps() === null ? 0 : self.noOfAncCheckUps();
+        });
         return self;
     };
 

--- a/custom/aaa/tasks.py
+++ b/custom/aaa/tasks.py
@@ -150,7 +150,6 @@ def prepare_export_reports(domain, selected_date, next_month_start, selected_loc
             ('noOfAncCheckUps', 'No. Of ANC Check-Ups'),
         )
         data = PregnantWomanQueryHelper.list(domain, selected_date, location_filters, sort_column)
-        data = PregnantWomanQueryHelper.update_list(data)
 
     export_columns = [col[1] for col in columns]
 

--- a/custom/aaa/tests/test_eligible_couple_query_helper.py
+++ b/custom/aaa/tests/test_eligible_couple_query_helper.py
@@ -65,7 +65,7 @@ class TestEligibleCoupleBeneficiarySections(TestCase):
                 'maleChildrenAlive': 0,
                 'femaleChildrenAlive': 0,
                 'familyPlaningMethod': 'condom',
-                'familyPlanningMethodDate': '2019-03-01',
+                'familyPlanningMethodDate': date(2019, 3, 1),
                 'ashaVisit': 'N/A',
                 'previousFamilyPlanningMethod': 'iud',
                 'preferredFamilyPlaningMethod': 'N/A',

--- a/custom/aaa/views.py
+++ b/custom/aaa/views.py
@@ -196,26 +196,35 @@ class UnifiedBeneficiaryReportAPI(View):
         sort_column_dir = self.request.POST.get('sortColumnDir', 'asc')
 
         location_filters = build_location_filters(selected_location, selected_ministry, with_child=False)
-
+        sort_column_with_dir = sort_column
         if sort_column_dir == 'desc':
-            sort_column = '-' + sort_column
+            sort_column_with_dir = '-' + sort_column
         data = []
         if beneficiary_type == 'child':
-            data = ChildQueryHelper.list(request.domain, next_month_start, location_filters, sort_column)
+            data = ChildQueryHelper.list(request.domain, next_month_start, location_filters, sort_column_with_dir)
         elif beneficiary_type == 'eligible_couple':
-            data = EligibleCoupleQueryHelper.list(request.domain, selected_date, location_filters, sort_column)
+            data = EligibleCoupleQueryHelper.list(
+                request.domain,
+                selected_date,
+                location_filters,
+                sort_column_with_dir
+            )
         elif beneficiary_type == 'pregnant_women':
-            data = PregnantWomanQueryHelper.list(request.domain, selected_date, location_filters, sort_column)
+            sort_column_with_dir = '"%s" %s' % (sort_column, sort_column_dir)
+            data = PregnantWomanQueryHelper.list(
+                request.domain,
+                selected_date,
+                location_filters,
+                sort_column_with_dir
+            )
         if data:
-            number_of_data = data.count()
+            number_of_data = len(data)
             data = data[start:start + length]
         else:
             number_of_data = 0
         if beneficiary_type == 'eligible_couple':
             month_end = date(selected_year, selected_month, 1) + relativedelta(months=1) - relativedelta(days=1)
             data = EligibleCoupleQueryHelper.update_list(data, month_end)
-        elif beneficiary_type == 'pregnant_women':
-            data = PregnantWomanQueryHelper.update_list(data)
         data = list(data)
         return JsonResponse(data={
             'rows': data,

